File: clients/node/conformance/runner.ts
// The conformance runner: it executes each scenario through the client's public
// API and compares the answer by the rules in conformance/README.md.

import { ConnectError } from "@connectrpc/connect";
import { readFileSync } from "node:fs";
import { isAbsolute, join, normalize, resolve as resolvePath, sep } from "node:path";
import type { DescMessage, Message } from "@bufbuild/protobuf";

import { SysMLService } from "../src/generated/sysml_pb.js";
import type { Connection } from "../src/core/connection.js";
import { Model } from "../src/core/model.js";
import { statusName } from "../src/core/status.js";
import { check, render } from "./compare.js";
import { byCodeUnit } from "./order.js";
import { MODEL_HASH_PLACEHOLDER, Normalizer } from "./normalize.js";
import { Literal, methodOf, type Expect, type Scenario, type ScenarioModel } from "./scenarios.js";

/** The RPCs v1 of this client covers. Everything else is a stated skip. */
export const COVERED_RPCS = ["GetServerInfo", "ParseFile", "GetSymbol", "Evaluate", "Instantiate"] as const;

/** One scenario's outcome. The shape cmd/conformance writes. */
export interface Result {
  id: string;
  outcome: "pass" | "fail" | "skip" | "error";
  rpc: string;
  reason?: string;
  failures?: string[];
  status: string;
  duration_ms: number;
}

/** One protocol's results. */
export interface Summary {
  protocol: string;
  service: string;
  capabilities: string[];
  total: number;
  passed: number;
  failed: number;
  skipped: number;
  errored: number;
  results: Result[];
}

/** The whole run, across protocols. */
export interface Report {
  service: string;
  total: number;
  passed: number;
  failed: number;
  skipped: number;
  errored: number;
  protocols: Summary[];
}

/** A deliberate corruption of a response, used to prove the runner is not vacuous. */
export type Mutation = (method: string, response: Message) => void;

/** How a runner reports and what it is allowed to do to the answers. */
export interface RunnerOptions {
  /** Directory holding the suite's fixtures. */
  fixtures: string;
  /** How the report names the service under test. */
  service: string;
  /** Protocol name this runner's connection speaks, for the report. */
  protocol: string;
  verbose?: boolean;
  log?: (line: string) => void;
  mutate?: Mutation;
}

const FIXTURE_REFERENCE = /^\$\{fixture:([^}]+)\}$/;

/** A call the client made: the response message and the schema to read it by. */
interface Answer {
  schema: DescMessage;
  message: Message;
}

/** What a scenario's call did, or why this client cannot make it. */
type Attempt =
  | { kind: "skip"; reason: string }
  | { kind: "done"; status: string; message: string; answer?: Answer };

/** Runs the suite over one connection. */
export class Runner {
  private readonly captured = new Map<string, Message>();
  private readonly hashes = new Map<string, string>();
  private readonly parsed = new Map<string, Model>();
  private capabilities: string[] = [];

  constructor(
    private readonly connection: Connection,
    private readonly options: RunnerOptions,
  ) {}

  /**
   * The response tap the runner's connection must install: comparison needs the
   * message on the wire, which the ergonomic API deliberately does not return.
   */
  tap(): (event: { method: string; response: unknown }) => void {
    return ({ method, response }) => {
      if (isMessage(response)) {
        this.options.mutate?.(method, response);
        this.captured.set(method, response);
      }
    };
  }

  /** Reads the capabilities the scenarios gate on. */
  async readCapabilities(): Promise<void> {
    const info = await this.connection.serverInfo();
    this.capabilities = [...info.capabilities].sort(byCodeUnit);
  }

  /** Runs every scenario whose id matches `filter`, reporting as it goes. */
  async runAll(scenarios: Scenario[], filter?: RegExp): Promise<Summary> {
    const summary: Summary = {
      protocol: this.options.protocol,
      service: this.options.service,
      capabilities: this.capabilities,
      total: 0,
      passed: 0,
      failed: 0,
      skipped: 0,
      errored: 0,
      results: [],
    };
    for (const scenario of scenarios) {
      if (filter !== undefined && !filter.test(scenario.id)) {
        continue;
      }
      const result = await this.run(scenario);
      summary.results.push(result);
      summary.total += 1;
      switch (result.outcome) {
        case "pass":
          summary.passed += 1;
          break;
        case "fail":
          summary.failed += 1;
          break;
        case "skip":
          summary.skipped += 1;
          break;
        default:
          summary.errored += 1;
      }
      this.report(result);
    }
    this.log(
      `\n[${summary.protocol}] ${summary.total} scenarios: ${summary.passed} passed, ` +
        `${summary.failed} failed, ${summary.skipped} skipped, ${summary.errored} in error`,
    );
    return summary;
  }

  /** Runs one scenario: parse the model it names, make the call, compare. */
  async run(scenario: Scenario): Promise<Result> {
    const started = performance.now();
    const rpc = methodOf(scenario);
    const result: Result = { id: scenario.id, outcome: "pass", rpc, status: "OK", duration_ms: 0 };
    const finish = (): Result => {
      result.duration_ms = Math.round((performance.now() - started) * 1000) / 1000;
      return result;
    };

    let expect: Expect = scenario.expect ?? {};
    const missing = (scenario.requires_capabilities ?? []).filter(
      (capability) => !this.capabilities.includes(capability),
    );
    if (missing.length > 0) {
      if (scenario.expect_without_capability === undefined) {
        result.outcome = "skip";
        result.status = "-";
        result.reason = `the service does not report ${missing.join(", ")}`;
        return finish();
      }
      expect = scenario.expect_without_capability;
      result.reason = `the service does not report ${missing.join(", ")}, so the without-capability expectation applies`;
    }

    if (scenario.model?.fixtures !== undefined) {
      result.outcome = "skip";
      result.status = "-";
      result.reason = "v1 of this client parses one document at a time, not a model of several";
      return finish();
    }

    let modelHash = "";
    let request: Record<string, unknown>;
    try {
      if (scenario.model !== undefined) {
        modelHash = await this.modelHash(scenario.model);
      }
      request = this.resolve(scenario.request ?? {}, modelHash) as Record<string, unknown>;
    } catch (error) {
      finish();
      return errored(result, error);
    }

    this.captured.clear();
    let attempt: Attempt;
    try {
      attempt = await this.attempt(rpc, request, modelHash, scenario);
    } catch (error) {
      finish();
      return errored(result, error);
    }
    if (attempt.kind === "skip") {
      result.outcome = "skip";
      result.status = "-";
      result.reason = attempt.reason;
      return finish();
    }

    const wantStatus = expect.status ?? "OK";
    result.status = attempt.status;
    if (attempt.status !== "OK") {
      if (attempt.status.toLowerCase() !== wantStatus.toLowerCase()) {
        result.outcome = "fail";
        result.failures = [`status: ${attempt.status} (${attempt.message}), want ${wantStatus}`];
        return finish();
      }
      const wantMessage = expect.status_message_contains;
      if (wantMessage !== undefined && !attempt.message.includes(wantMessage)) {
        result.outcome = "fail";
        result.failures = [
          `status message ${JSON.stringify(attempt.message)} does not contain ${JSON.stringify(wantMessage)}`,
        ];
      }
      return finish();
    }
    if (wantStatus !== "OK") {
      result.outcome = "fail";
      result.failures = [`the call succeeded, want status ${wantStatus}`];
      return finish();
    }
    if (attempt.answer === undefined) {
      finish();
      return errored(result, new Error(`the client made no ${rpc} call the runner could compare`));
    }

    const normalized = new Normalizer(modelHash).normalize(attempt.answer.schema, attempt.answer.message);
    if (this.options.verbose === true) {
      this.log(`       ${render(normalized)}`);
    }
    const failures = check(expect, normalized);
    if (failures.length > 0) {
      result.outcome = "fail";
      result.failures = failures;
    }
    return finish();
  }

  /** Makes the scenario's call through the public API, or says why it cannot. */
  private async attempt(
    rpc: string,
    request: Record<string, unknown>,
    modelHash: string,
    scenario: Scenario,
  ): Promise<Attempt> {
    if (!COVERED_RPCS.includes(rpc as (typeof COVERED_RPCS)[number])) {
      return { kind: "skip", reason: `v1 of this client does not cover ${rpc}` };
    }
    const unsupported = this.unsupported(rpc, request);
    if (unsupported !== undefined) {
      return { kind: "skip", reason: unsupported };
    }
    try {
      await this.call(rpc, request, modelHash, scenario);
    } catch (error) {
      const connectError = asConnectError(error);
      if (connectError !== undefined) {
        return { kind: "done", status: statusName(connectError.code), message: connectError.rawMessage };
      }
      // The API raises a failure the service reported in a successful answer;
      // the answer itself is what the scenario compares.
      if (!this.captured.has(rpc)) {
        throw error;
      }
    }
    const message = this.captured.get(rpc);
    return {
      kind: "done",
      status: "OK",
      message: "",
      ...(message === undefined ? {} : { answer: { schema: outputSchema(rpc), message } }),
    };
  }

  /** Why the public API cannot express this request, when it cannot. */
  private unsupported(rpc: string, request: Record<string, unknown>): string | undefined {
    if (rpc !== "ParseFile") {
      return undefined;
    }
    if (typeof request["content"] !== "string" && typeof request["file_path"] !== "string") {
      return "the client's load()/loads() always name a source, so a request naming neither cannot be made through it";
    }
    return undefined;
  }

  private async call(
    rpc: string,
    request: Record<string, unknown>,
    modelHash: string,
    scenario: Scenario,
  ): Promise<void> {
    switch (rpc) {
      case "GetServerInfo":
        await this.connection.serverInfo();
        return;
      case "ParseFile": {
        const options = {
          ...(typeof request["language"] === "string" ? { language: request["language"] } : {}),
          ...(request["strict_conformance"] === true ? { strict: true } : {}),
        };
        const content = request["content"];
        if (typeof content === "string") {
          await this.connection.loads(content, options);
          return;
        }
        await this.connection.load(String(request["file_path"]), options);
        return;
      }
      case "GetSymbol":
        await this.model(request, modelHash, scenario).symbolById(String(request["symbol_id"]));
        return;
      case "Evaluate": {
        const model = this.model(request, modelHash, scenario);
        await model.eval(String(request["expression"]), {
          ...(typeof request["context_symbol_id"] === "string" ? { context: request["context_symbol_id"] } : {}),
          ...(typeof request["subject_symbol_id"] === "string" ? { subject: request["subject_symbol_id"] } : {}),
        });
        return;
      }
      case "Instantiate":
        await this.model(request, modelHash, scenario).instantiate(String(request["symbol_id"]));
        return;
      default:
        throw new Error(`the runner covers ${COVERED_RPCS.join(", ")}, not ${rpc}`);
    }
  }

  /**
   * The model a scenario's call addresses: the one parsed from its fixture, or a
   * hash adopted as written, which is how "no-such-model" reaches the service.
   */
  private model(request: Record<string, unknown>, modelHash: string, scenario: Scenario): Model {
    const named = request["model_hash"];
    const hash = typeof named === "string" ? named : modelHash;
    if (hash === "") {
      throw new Error(`${scenario.id} names no model hash`);
    }
    return this.parsed.get(hash) ?? this.connection.model(hash);
  }

  /** Parses a scenario's fixture once per run and remembers the hash. */
  private async modelHash(model: ScenarioModel): Promise<string> {
    if (model.fixture === undefined) {
      throw new Error("a model names no fixture");
    }
    const key = `${model.fixture}|${model.language ?? ""}|${String(model.strict_conformance ?? false)}`;
    const known = this.hashes.get(key);
    if (known !== undefined) {
      return known;
    }
    const source = this.fixture(model.fixture);
    const parsed = await this.connection.loads(source, {
      ...(model.language === undefined ? {} : { language: model.language }),
      ...(model.strict_conformance === undefined ? {} : { strict: model.strict_conformance }),
    });
    const failure = parsed.diagnostics.find((diagnostic) => diagnostic.severity === "error");
    if (failure !== undefined) {
      throw new Error(`fixture ${JSON.stringify(model.fixture)} does not parse clean: ${failure.message}`);
    }
    if (parsed.hash === "") {
      throw new Error(`parsing fixture ${JSON.stringify(model.fixture)} returned no model hash`);
    }
    this.hashes.set(key, parsed.hash);
    this.parsed.set(parsed.hash, parsed);
    return parsed.hash;
  }

  /** Replaces "${model_hash}" and "${fixture:<name>}" in a request. */
  private resolve(tree: unknown, modelHash: string): unknown {
    if (Array.isArray(tree)) {
      return tree.map((item) => this.resolve(item, modelHash));
    }
    if (typeof tree === "object" && tree !== null && !(tree instanceof Literal)) {
      const out: Record<string, unknown> = {};
      for (const [key, value] of Object.entries(tree)) {
        out[key] = this.resolve(value, modelHash);
      }
      return out;
    }
    if (typeof tree === "string") {
      if (tree === MODEL_HASH_PLACEHOLDER) {
        if (modelHash === "") {
          throw new Error(`the request names ${MODEL_HASH_PLACEHOLDER} but the scenario declares no model`);
        }
        return modelHash;
      }
      const reference = FIXTURE_REFERENCE.exec(tree);
      if (reference !== null) {
        return this.fixture(reference[1]);
      }
    }
    return tree;
  }

  /** Reads a fixture's source, refusing a name that leaves the fixtures directory. */
  private fixture(name: string): string {
    const fixtures = resolvePath(this.options.fixtures);
    const path = isAbsolute(name) ? normalize(name) : resolvePath(join(fixtures, normalize(name)));
    if (!path.startsWith(fixtures + sep)) {
      throw new Error(`fixture ${JSON.stringify(name)} is outside ${fixtures}`);
    }
    return readFileSync(path, "utf8");
  }

  private report(result: Result): void {
    const marks = { pass: "PASS", fail: "FAIL", skip: "SKIP", error: "ERR " };
    this.log(`${marks[result.outcome]} ${result.id.padEnd(46)} ${result.status}`);
    if (result.reason !== undefined) {
      this.log(`       ${result.reason}`);
    }
    for (const failure of result.failures ?? []) {
      this.log(`       ${failure}`);
    }
  }

  private log(line: string): void {
    (this.options.log ?? console.log)(line);
  }
}

/** Adds one protocol's results to a report. */
export function accumulate(report: Report, summary: Summary): void {
  report.protocols.push(summary);
  report.total += summary.total;
  report.passed += summary.passed;
  report.failed += summary.failed;
  report.skipped += summary.skipped;
  report.errored += summary.errored;
}

/** The response schema of an RPC, which the normalizer reads the answer by. */
function outputSchema(rpc: string): DescMessage {
  const method = Object.values(SysMLService.method).find((candidate) => candidate.name === rpc);
  if (method === undefined) {
    throw new Error(`${SysMLService.typeName} has no RPC ${JSON.stringify(rpc)}`);
  }
  return method.output;
}

/** The Connect error an exception carries, whether raised directly or wrapped. */
function asConnectError(error: unknown): ConnectError | undefined {
  if (error instanceof ConnectError) {
    return error;
  }
  if (error instanceof Error && error.cause instanceof ConnectError) {
    return error.cause;
  }
  return undefined;
}

function errored(result: Result, error: unknown): Result {
  result.outcome = "error";
  result.status = "-";
  result.reason = error instanceof Error ? error.message : String(error);
  return result;
}

function isMessage(value: unknown): value is Message {
  return typeof value === "object" && value !== null && "$typeName" in value;
}

File: client/java/opensysml-conformance/src/main/java/org/openmbee/opensysml/conformance/Runner.java
package org.openmbee.opensysml.conformance;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonPrimitive;
import com.google.protobuf.InvalidProtocolBufferException;
import com.google.protobuf.Message;
import com.google.protobuf.util.JsonFormat;
import org.openmbee.opensysml.Connection;
import org.openmbee.opensysml.Diagnostic;
import org.openmbee.opensysml.Model;
import org.openmbee.opensysml.ModelException;
import org.openmbee.opensysml.OpenSysMLException;
import org.openmbee.opensysml.ParseOptions;
import java.io.IOException;
import java.io.PrintStream;
import java.io.UncheckedIOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.TreeSet;
import java.util.function.UnaryOperator;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/** Runs scenarios against one service through one encoding of the Connect protocol. */
final class Runner {

  private static final String DETAIL_FORMAT = "       %s%n";
  private static final Pattern FIXTURE_REFERENCE = Pattern.compile("^\\$\\{fixture:([^}]+)}$");

  private final String protocol;
  private final Connection connection;
  private final Api api;
  private final Path fixtures;
  private final List<String> capabilities;
  private final Map<Scenario.Fixture, String> models = new HashMap<>();
  private final PrintStream out;
  private final boolean verbose;
  private final UnaryOperator<Message> mutation;

  Runner(
      String protocol,
      Connection connection,
      Path fixtures,
      PrintStream out,
      boolean verbose,
      UnaryOperator<Message> mutation) {
    this.protocol = protocol;
    this.connection = connection;
    this.api = new Api(connection);
    this.fixtures = fixtures;
    this.capabilities = List.copyOf(new TreeSet<>(connection.capabilities().names()));
    this.out = out;
    this.verbose = verbose;
    this.mutation = mutation;
  }

  /**
   * Runs every scenario whose id matches, reporting as it goes.
   *
   * @param scenarios the scenarios to run
   * @param filter an id pattern, or empty for all of them
   * @return this protocol's summary
   */
  Report.Summary runAll(List<Scenario> scenarios, Optional<Pattern> filter) {
    Report.Summary summary = new Report.Summary(protocol, connection.address(), capabilities);
    for (Scenario scenario : scenarios) {
      if (filter.isPresent() && !filter.get().matcher(scenario.id()).find()) {
        continue;
      }
      Report.Result result = run(scenario);
      summary.results.add(result);
      summary.total++;
      switch (result.outcome) {
        case "pass" -> summary.passed++;
        case "fail" -> summary.failed++;
        case "skip" -> summary.skipped++;
        default -> summary.errored++;
      }
      report(result);
    }
    out.printf(
        "%n[%s] %d scenarios: %d passed, %d failed, %d skipped, %d in error%n",
        protocol, summary.total, summary.passed, summary.failed, summary.skipped, summary.errored);
    return summary;
  }

  private void report(Report.Result result) {
    String mark =
        switch (result.outcome) {
          case "pass" -> "PASS";
          case "fail" -> "FAIL";
          case "skip" -> "SKIP";
          default -> "ERR ";
        };
    out.printf("%s %-46s %s%n", mark, result.id, result.status);
    if (result.reason != null) {
      out.printf(DETAIL_FORMAT, result.reason);
    }
    if (result.failures != null) {
      result.failures.forEach(failure -> out.printf(DETAIL_FORMAT, failure));
    }
  }

  /** Runs one scenario: parses the model it names, makes the call, compares the answer. */
  private Report.Result run(Scenario scenario) {
    long started = System.nanoTime();
    Report.Result result = new Report.Result(scenario.id(), scenario.method());
    try {
      execute(scenario, result);
    } catch (RuntimeException e) {
      errored(result, e.toString());
    }
    result.durationMs = (System.nanoTime() - started) / 1_000_000.0;
    return result;
  }

  private void execute(Scenario scenario, Report.Result result) {
    if (!Api.declared(scenario.method())) {
      errored(result, "sysml.SysMLService has no RPC " + scenario.method());
      return;
    }

    if (!Api.COVERED.contains(scenario.method())) {
      result.outcome = "skip";
      result.status = "-";
      result.reason = "the public API does not cover " + scenario.method();
      return;
    }

    Scenario.Expect expect = scenario.expect();
    List<String> missing = missingCapabilities(scenario);
    if (!missing.isEmpty()) {
      Optional<Scenario.Expect> withoutCapability = scenario.expectWithoutCapability();
      if (withoutCapability.isEmpty()) {
        result.outcome = "skip";
        result.status = "-";
        result.reason = "the service does not report " + String.join(", ", missing);
        return;
      }
      expect = withoutCapability.get();
      result.reason =
          "the service does not report "
              + String.join(", ", missing)
              + ", so the without-capability expectation applies";
    }

    String modelHash = "";
    Optional<Scenario.Fixture> model = scenario.model();
    if (model.isPresent()) {
      try {
        modelHash = modelHash(model.get());
      } catch (OpenSysMLException | IllegalStateException e) {
        errored(result, e.getMessage());
        return;
      }
    }

    Message request;
    try {
      request = request(scenario, modelHash);
    } catch (RuntimeException e) {
      errored(result, e.getMessage());
      return;
    }

    Api.Answer answer = api.call(scenario.method(), request);
    if (answer instanceof Api.Answer.Unsupported unsupported) {
      result.outcome = "skip";
      result.status = "-";
      result.reason = unsupported.reason();
      return;
    }
    if (answer instanceof Api.Answer.Refused refused) {
      checkRefusal(expect, refused, result);
      return;
    }

    Message response = mutation.apply(((Api.Answer.Answered) answer).response());
    if (!expect.wantStatus().equals("OK")) {
      result.outcome = "fail";
      result.failures = List.of("the call succeeded, want status " + expect.wantStatus());
      return;
    }
    Map<String, Object> normalized = new Normalizer(modelHash).normalize(response);
    if (verbose) {
      out.printf(DETAIL_FORMAT, Checks.render(normalized));
    }
    List<String> failures = Checks.check(expect, normalized);
    if (!failures.isEmpty()) {
      result.outcome = "fail";
      result.failures = failures;
    }
  }

  private void checkRefusal(
      Scenario.Expect expect, Api.Answer.Refused refused, Report.Result result) {
    result.status = refused.status();
    if (!expect.wantStatus().equalsIgnoreCase(refused.status())) {
      result.outcome = "fail";
      result.failures =
          List.of(
              "status: "
                  + refused.status()
                  + " ("
                  + refused.message()
                  + "), want "
                  + expect.wantStatus());
      return;
    }
    if (!expect.statusMessageContains().isEmpty()
        && !refused.message().contains(expect.statusMessageContains())) {
      result.outcome = "fail";
      result.failures =
          List.of(
              "status message \""
                  + refused.message()
                  + "\" does not contain \""
                  + expect.statusMessageContains()
                  + "\"");
    }
  }

  /** The capabilities a scenario needs that the service does not report. */
  private List<String> missingCapabilities(Scenario scenario) {
    List<String> missing = new ArrayList<>();
    for (String capability : scenario.requiresCapabilities()) {
      if (!capabilities.contains(capability)) {
        missing.add(capability);
      }
    }
    return missing;
  }

  /** Builds the call's message from the scenario's protobuf JSON, with the placeholders resolved. */
  private Message request(Scenario scenario, String modelHash) {
    Message.Builder builder = Api.request(scenario.method());
    if (scenario.request().isEmpty()) {
      return builder.build();
    }
    JsonElement resolved = resolve(scenario.request(), modelHash);
    try {
      JsonFormat.parser().merge(resolved.toString(), builder);
    } catch (InvalidProtocolBufferException e) {
      throw new IllegalArgumentException(
          "request does not fit " + builder.getDescriptorForType().getFullName() + ": " + e.getMessage(),
          e);
    }
    return builder.build();
  }

  /** Replaces {@code ${model_hash}} and {@code ${fixture:<path>}} in a request. */
  private JsonElement resolve(JsonElement tree, String modelHash) {
    if (tree instanceof JsonObject object) {
      JsonObject resolved = new JsonObject();
      object.entrySet().forEach(entry -> resolved.add(entry.getKey(), resolve(entry.getValue(), modelHash)));
      return resolved;
    }
    if (tree instanceof JsonArray array) {
      JsonArray resolved = new JsonArray();
      array.forEach(item -> resolved.add(resolve(item, modelHash)));
      return resolved;
    }
    if (tree instanceof JsonPrimitive primitive && primitive.isString()) {
      String text = primitive.getAsString();
      if (text.equals(Normalizer.MODEL_HASH)) {
        if (modelHash.isEmpty()) {
          throw new IllegalArgumentException(
              "the request names " + Normalizer.MODEL_HASH + " but the scenario declares no model");
        }
        return new JsonPrimitive(modelHash);
      }
      Matcher reference = FIXTURE_REFERENCE.matcher(text);
      if (reference.matches()) {
        return new JsonPrimitive(fixture(reference.group(1)));
      }
    }
    return tree;
  }

  /** A fixture's source, refusing a name that reaches outside the fixtures directory. */
  private String fixture(String name) {
    Path path = fixtures.resolve(name).normalize();
    if (!path.startsWith(fixtures.normalize())) {
      throw new IllegalArgumentException("fixture " + name + " is outside " + fixtures);
    }
    try {
      return Files.readString(path, StandardCharsets.UTF_8);
    } catch (IOException e) {
      throw new UncheckedIOException("reading fixture " + name, e);
    }
  }

  /** Parses a scenario's model once per run; a model that does not parse clean is a suite error. */
  private String modelHash(Scenario.Fixture fixture) {
    String cached = models.get(fixture);
    if (cached != null) {
      return cached;
    }
    ParseOptions options =
        new ParseOptions(
            org.openmbee.opensysml.Language.fromWireName(fixture.language()), fixture.strictConformance());
    Model model;
    try {
      if (!fixture.fixtures().isEmpty()) {
        List<org.openmbee.opensysml.SourceDocument> documents =
            fixture.fixtures().stream()
                .map(name -> org.openmbee.opensysml.SourceDocument.inline(name, fixture(name)))
                .toList();
        model = connection.parseSources(documents, options);
      } else {
        model = connection.parse(fixture(fixture.fixture()), options);
      }
    } catch (ModelException e) {
      throw new IllegalStateException("parsing fixture " + fixture.fixture() + ": " + e.getMessage(), e);
    }
    for (Diagnostic diagnostic : model.parseDiagnostics()) {
      if (diagnostic.severity() == Diagnostic.Severity.ERROR) {
        throw new IllegalStateException(
            "fixture " + fixture.fixture() + " does not parse clean: " + diagnostic.message());
      }
    }
    if (model.hash().isEmpty()) {
      throw new IllegalStateException("parsing fixture " + fixture.fixture() + " returned no model hash");
    }
    models.put(fixture, model.hash());
    return model.hash();
  }

  private static void errored(Report.Result result, String reason) {
    result.outcome = "error";
    result.status = "-";
    result.reason = reason;
  }
}
